## Loading packages

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.impute import SimpleImputer   # Instead of sklearn.preprocessing import Imputer
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectFromModel
from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier

pd.set_option('display.max_columns', 100)

## Loading data

In [4]:
train = pd.read_csv('train.csv')
test = pd. read_csv('test.csv')

## Data at first sight

Here is an excerpt of the data description for the competition:  
  
- Features that belongs to **similar groupings are tagged** as such in the feature names (e.g., ind, reg, car, calc)  
  
- **Feature names include the postfix** bin to indicate binary features and cat to indicate categorical features.  
  
- Features **without these designations are either continuous or ordinal**.  
  
- **Values of -1 indicate that the feature was missing** from the observation.  
  
- The **target columns** signifies whether or not a claim was filed for that policy holder. ==> Binary Classification Problem.

Let's have a quick look at the first and last rows to confirm all of this.

In [5]:
train.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0              0              0              1              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              1              0              0              0   
4              0              1              0              0              0   

   ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  \
0              0              0              0              0          0   
1              0              0              0              0          0   
2              0              0              0              0          0   
3              0              0              0              0          0   
4              0              0              0              0          0   

   ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  \
0         11              0              1              0        0.7   
1          3              0              0              1        0.8   
2         12              1              0              0        0.0   
3          8              1              0              0        0.9   
4          9              1              0              0        0.7   

   ps_reg_02  ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  \
0        0.2   0.718070             10              1             -1   
1        0.4   0.766078             11              1             -1   
2        0.0  -1.000000              7              1             -1   
3        0.2   0.580948              7              1              0   
4        0.6   0.840759             11              1             -1   

   ps_car_04_cat  ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  \
0              0              1              4              1              0   
1              0             -1             11              1              1   
2              0             -1             14              1              1   
3              0              1             11              1              1   
4              0             -1             14              1              1   

   ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  \
0              0              1             12          2   0.400000   
1              2              1             19          3   0.316228   
2              2              1             60          1   0.316228   
3              3              1            104          1   0.374166   
4              2              1             82          3   0.316070   

   ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
0   0.883679   0.370810   3.605551         0.6         0.5         0.2   
1   0.618817   0.388716   2.449490         0.3         0.1         0.3   
2   0.641586   0.347275   3.316625         0.5         0.7         0.1   
3   0.542949   0.294958   2.000000         0.6         0.9         0.1   
4   0.565832   0.365103   2.000000         0.4         0.6         0.0   

   ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
0           3           1          10           1          10           1   
1           2           1           9           5           8           1   
2           2           2           9           1           8           2   
3           2           4           7           1           8           4   
4           2           2           6   

In [6]:
train.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
595207              0        0.5        0.3   0.692820             10   
595208              0        0.9        0.7   1.382027              9   
595209              0        0.9        0.2   0.659071              7   
595210              0        0.9        0.4   0.698212             11   
595211              0        0.1        0.2  -1.000000              7   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
595207              1             -1              0              1   
595208              1             -1              0             -1   
595209              1             -1              0             -1   
595210              1             -1              0             -1   
595211              0             -1              0             -1   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
595207              1              1              1              0   
595208             15              0              0              2   
595209              1              1              1              2   
595210             11              1              1              2   
595211              0              1              0              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
595207              1             31          3   0.374166   0.684631   
595208              1             63          2   0.387298   0.972145   
595209              1             31          3   0.397492   0.596373   
595210              1            101          3   0.374166   0.764434   
595211              1             34          2   0.400000   0.932649   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
595207   0.385487   2.645751         0.4         0.5         0.3           3   
595208  -1.000000   3.605551         0.2         0.2         0.0           2   
595209   0.398748   1.732051         0.4         0.0         0.3           3   
595210   0.384968   3.162278         0.0         0.7         0.0           4   
595211   0.378021   3.741657         0.4         0.0         0.5           2   

        ps_calc_05  ps_calc_06  

We indeed see the following:  
  
- binary variables
- categorical variables of which the category values are integers
- other variables with integer or float values
- other variables with -1 representing missing values
- the target variables and id variable

Let's look at the number of rows and columns in the train data.

In [7]:
train.shape

(595212, 59)

We have 59 variables and 595,212 rows.  
Let's see if we have the same number of variables in the test data.  

In [8]:
test.shape

(892816, 58)

We are missing one variable in the test data, but this is the target variable.

Let's see if there are duplicate rows in the training data.

In [11]:
train.drop_duplicates()
train.shape

(595212, 59)

No duplicate rows, so that's fine.

Now, let's investigate how many variables of each type we have.  
  
So later on we can create dummy variables for the 14 categorical variables.  
The *bin* variables are already binary and do not need dummification. 

In [13]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595212 entries, 0 to 595211
Data columns (total 59 columns):
id                595212 non-null int64
target            595212 non-null int64
ps_ind_01         595212 non-null int64
ps_ind_02_cat     595212 non-null int64
ps_ind_03         595212 non-null int64
ps_ind_04_cat     595212 non-null int64
ps_ind_05_cat     595212 non-null int64
ps_ind_06_bin     595212 non-null int64
ps_ind_07_bin     595212 non-null int64
ps_ind_08_bin     595212 non-null int64
ps_ind_09_bin     595212 non-null int64
ps_ind_10_bin     595212 non-null int64
ps_ind_11_bin     595212 non-null int64
ps_ind_12_bin     595212 non-null int64
ps_ind_13_bin     595212 non-null int64
ps_ind_14         595212 non-null int64
ps_ind_15         595212 non-null int64
ps_ind_16_bin     595212 non-null int64
ps_ind_17_bin     595212 non-null int64
ps_ind_18_bin     595212 non-null int64
ps_reg_01         595212 non-null float64
ps_reg_02         595212 non-null float64
ps_re

Again, with the info() method we see that the data type is integer or float. No null values are present in the data set. That's normal because missing values are raplaced by -1. We will look into that later.

## Metadata  
  
To facilitate the data management, we'll store meta-information about the variables in a DataFrame. This will be helpful when we want to select specific variables for analysis, visualization, modeling, etc.  
  
Concretely we will store:  
  
- **role**: input, ID, target  
- **level**: nominal, interval, ordinal, binary
- **keep**: True or False
- **dtype**: int, float, str

In [23]:
data = []

for f in train.columns:
    # Defining the role
    if f == 'target':
        role = 'target'
    elif f == 'id':
        role = 'id'
    else:
        role = 'input'
        
    # Defining the level
    if 'bin' in f or f == 'target':
        level = 'binary'
    elif 'cat' in f or f == 'id':
        level = 'nominal'
    elif train[f].dtype == float:
        level = 'interval'
    else:
        level = 'ordinal'
        
    # Initialize keep to True for all variables except for id
    keep = True
    if f == 'id':
        keep = False
        
    # Defining the data type
    dtype = train[f].dtype
    
    # Creating a Dict that contains all the metadata for the variable.
    f_dict = {'varname': f,
             'role': role,
             'level': level,
             'keep': keep,
             'dtype': dtype}
    data.append(f_dict)
    
meta = pd.DataFrame(data, columns=['varname', 'role','level','keep','dtype'])
meta.set_index('varname', inplace=True)

In [24]:
meta

role     level   keep    dtype
varname                                         
id                  id   nominal  False    int64
target          target    binary   True    int64
ps_ind_01        input   ordinal   True    int64
ps_ind_02_cat    input   nominal   True    int64
ps_ind_03        input   ordinal   True    int64
ps_ind_04_cat    input   nominal   True    int64
ps_ind_05_cat    input   nominal   True    int64
ps_ind_06_bin    input    binary   True    int64
ps_ind_07_bin    input    binary   True    int64
ps_ind_08_bin    input    binary   True    int64
ps_ind_09_bin    input    binary   True    int64
ps_ind_10_bin    input    binary   True    int64
ps_ind_11_bin    input    binary   True    int64
ps_ind_12_bin    input    binary   True    int64
ps_ind_13_bin    input    binary   True    int64
ps_ind_14        input   ordinal   True    int64
ps_ind_15        input   ordinal   True    int64
ps_ind_16_bin    input    binary   True    int64
ps_ind_17_bin    input    binary   True    int64
ps_ind_18_bin    input    binary   True    int64
ps_reg_01        input  interval   True  float64
ps_reg_02        input  interval   True  float64
ps_reg_03        input  interval   True  float64
ps_car_01_cat    input   nominal   True    int64
ps_car_02_cat    input   nominal   True    int64
ps_car_03_cat    input   nominal   True    int64
ps_car_04_cat    input   nominal   True    int64
ps_car_05_cat    input   nominal   True    int64
ps_car_06_cat    input   nominal   True    int64
ps_car_07_cat    input   nominal   True    int64
ps_car_08_cat    input   nominal   True    int64
ps_car_09_cat    input   nominal   True    int64
ps_car_10_cat    input   nominal   True    int64
ps_car_11_cat    input   nominal   True    int64
ps_car_11        input   ordinal   True    int64
ps_car_12        input  interval   True  float64
ps_car_13        input  interval   True  float64
ps_car_14        input  interval   True  float64
ps_car_15        input  interval   True  float64
ps_calc_01       input  interval   True  float64
ps_calc_02       input  interval   True  float64
ps_calc_03       input  interval   True  float64
ps_calc_04       input   ordinal   True    int64
ps_calc_05       input   ordinal   True    int64
ps_calc_06       input   ordinal   True    int64
ps_calc_07       input   ordinal   True    int64
ps_calc_08       input   ordinal   True    int64
ps_calc_09       input   ordinal   True    int64
ps_calc_10       input   ordinal   True    int64
ps_calc_11       input   ordinal   True    int64
ps_calc_12       input   ordinal   True    int64
ps_calc_13       input   ordinal   True    int64
ps_calc_14       input   ordinal   True    int64
ps_calc_15_bin   input    binary   True    int64
ps_calc_16_bin   input    binary   True    int64
ps_calc_17_bin   input    binary   True    int64
ps_calc_18_bin   input    binary   True    int64
ps_calc_19_bin   input    binary   True    int64
ps_calc_20_bin   input    binary   True    int64

Example to extract all nominal variables that are not dropped.

In [27]:
meta[(meta.level == 'nominal') & (meta.keep)].index

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_car_01_cat',
       'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat',
       'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat',
       'ps_car_10_cat', 'ps_car_11_cat'],
      dtype='object', name='varname')

Below the number of variables per role and level are displayed.

In [36]:
pd.DataFrame({'count': meta.groupby(['role','level'])['role'].count()}).reset_index()

role     level  count
0      id   nominal      1
1   input    binary     17
2   input  interval     10
3   input   nominal     14
4   input   ordinal     16
5  target    binary      1

## Descriptive statistics  
  
We can also apply the *describe* method on the dataframe. However, it doesn't make much sense to calculate the mean, std on categorical variables and the id variable. We will explore the categorical variables visually later.  
  
Thanks to our meta file we can easily select the variables on which we want to compute the descriptive statistics. To keep things clear, we'll do this per data type.

### Interval variables (continuous variables)

In [37]:
v = meta[(meta.level=='interval') & meta.keep].index
train[v].describe()

ps_reg_01      ps_reg_02      ps_reg_03      ps_car_12  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.610991       0.439184       0.551102       0.379945   
std         0.287643       0.404264       0.793506       0.058327   
min         0.000000       0.000000      -1.000000      -1.000000   
25%         0.400000       0.200000       0.525000       0.316228   
50%         0.700000       0.300000       0.720677       0.374166   
75%         0.900000       0.600000       1.000000       0.400000   
max         0.900000       1.800000       4.037945       1.264911   

           ps_car_13      ps_car_14      ps_car_15     ps_calc_01  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.813265       0.276256       3.065899       0.449756   
std         0.224588       0.357154       0.731366       0.287198   
min         0.250619      -1.000000       0.000000       0.000000   
25%         0.670867       0.333167       2.828427       0.200000   
50%         0.765811       0.368782       3.316625       0.500000   
75%         0.906190       0.396485       3.605551       0.700000   
max         3.720626       0.636396       3.741657       0.900000   

          ps_calc_02     ps_calc_03  
count  595212.000000  595212.000000  
mean        0.449589       0.449849  
std         0.286893       0.287153  
min         0.000000       0.000000  
25%         0.200000       0.200000  
50%         0.400000       0.500000  
75%         0.700000       0.700000  
max         0.900000       0.900000

#### reg variables  
- only ps_reg_03 has missing values
- the range differs between the variables. We could apply scaling (eg.StandardScaler), but it depends on the classifier we will want to use.  
  
#### car variables  
  
- ps_car_12 and ps_car_14 have missing values
- again, the range differs and we could apply scaling depending on our model.  
  
#### calc variables  
  
- no missing values
- This seems to be some kind of ratio as the maximum is 0.9
- all three \_calc variables have very similar distriutions

**Overall**, we can see that the range of the interval variables is rather small. Perhaps some transformation(eg.log) is already applied in order to anonymize the data?

### Ordinal variables